### PR TITLE
Loading CSV using Spanner JDBC Driver

### PR DESCRIPTION
**Pull request by kaylaleung**
- [x] **Tests** pass:   `mvn clean verify` **required**
- [x] **Lint**  passes: `mvn -P lint checkstyle:check` **required**

### CODE REVIEW

File: spanner/cloud-client/src/main/java/com/example/spanner/LoadCsvExample.java [hunk 1]
```diff
@@ -0,0 +1,250 @@
...
+      return SpannerDataTypes.DATE;
+    } else if (columnType.equalsIgnoreCase("TIMESTAMP")) {
+      return SpannerDataTypes.TIMESTAMP;
+    } else {
+      throw new IllegalArgumentException(
+          "Unrecognized or unsupported column data type: " + columnType);
+    }
+  }
+
+  public static void parseTableTypes(String tableId) throws SQLException {
```
Comment (dttran) [inline spanner/cloud-client/src/main/java/com/example/spanner/LoadCsvExample.java]: maybe change to tableName instead?

File: spanner/cloud-client/src/main/java/com/example/spanner/LoadCsvExample.java [hunk 2]
```diff
@@ -0,0 +1,250 @@
...
+      headers.add(tableHeader.getString("column_name"));
+    }
+  }
+
+  public static boolean isValidHeaders(CSVParser parser) {
+    List<String> allHeaders = parser.getHeaderNames();
+    for (int i = 0; i < allHeaders.size(); i++) {
+      if (!allHeaders.get(i).equals(headers.get(i))) {
+        System.err.println(
+            "Header " + allHeaders.get(i) + " does not match database table header " + headers
```
Comment (dttran) [inline spanner/cloud-client/src/main/java/com/example/spanner/LoadCsvExample.java]: Maybe change to "does not match database table column name" instead? A table doesn't have a header.

File: spanner/cloud-client/src/main/java/com/example/spanner/LoadCsvExample.java [hunk 3]
```diff
@@ -0,0 +1,250 @@
...
+      CSVParser parser = CSVParser.parse(in, parseFormat);
+
+      if (cmd.hasOption("h") && !isValidHeaders(parser)) {
+        return;
+      }
+
+      try {
+        loadData(parser, tableId);
+      } catch (SQLException e) {
+        System.err.println("Data could not be written to the database table.");
```
Comment (dttran) [inline spanner/cloud-client/src/main/java/com/example/spanner/LoadCsvExample.java]: Could you print out the exception so user know what the error is?

File: spanner/cloud-client/src/main/java/com/example/spanner/LoadCsvExample.java [hunk 4]
```diff
@@ -0,0 +1,250 @@
...
+    if (cmd.hasOption("h") &&  cmd.getOptionValue("h").equalsIgnoreCase("True")) {
+      parseFormat = parseFormat.withFirstRecordAsHeader();
+    }
+    return parseFormat;
+  }
+
+  public static void loadData(Iterable<CSVRecord> records, String tableId)
+      throws SQLException {
+    System.out.println("Writing data into table...");
+    List<Mutation> mutations = new ArrayList<>();
```
Comment (dttran) [inline spanner/cloud-client/src/main/java/com/example/spanner/LoadCsvExample.java:171]: Is this Mutation used in the JDBC driver too?

**Review by dttran — COMMENTED** [review]
No specific comment, but in general please follow the Java style guide in go/java-style.
Could you also add more code comments, and comment for every function?

File: spanner/cloud-client/src/main/java/com/example/spanner/LoadCsvExample.java [hunk 5]
```diff
@@ -0,0 +1,250 @@
...
+    for (CSVRecord record : records) {
+      WriteBuilder builder = Mutation.newInsertOrUpdateBuilder(tableId);
+      for (int i = 0; i < headers.size(); i++) {
```
Comment (dttran) [inline spanner/cloud-client/src/main/java/com/example/spanner/LoadCsvExample.java]: One more feature that you could add, which will help your COPY work is:
- If there is no header specified, assume the CSV file contains all columns of the table.
- If there is a header, the CSV can contains a subset of columns (other columns will have the NULL value).

Could you also add a unit test with null strings? To make sure that these values are written as NULL in Spanner.

File: spanner/cloud-client/src/main/java/com/example/spanner/LoadCsvExample.java (hunk 4 above)
Comment (kaylaleung) [reply spanner/cloud-client/src/main/java/com/example/spanner/LoadCsvExample.java:171]: Not entirely sure if I completely understand what you're asking, but this is the mutation that is passed into the `CloudSpannerJdbcConnection` write method [here](https://github.com/kaylaleung/java-docs-samples/pull/1/files/0b393b38463ac866aed7bf64af4c0593aae6cd45#diff-9c10a7320fc201516b9898e887d21f8d75ea1ccbbb9f1590fcdda512d338ae1dR192) so I think the answer is yes?

**Comment by kaylaleung** [comment]
I'm not sure what the best convention here is with regards to if I should also be loading my sample csv files used for testing into this PR as well?

File: spanner/cloud-client/src/main/java/com/example/spanner/LoadCsvExample.java (hunk 5 above)
Comment (kaylaleung) [reply spanner/cloud-client/src/main/java/com/example/spanner/LoadCsvExample.java]: Updated so that both those features are now added. Also added a test for parsing with the null strings.

Not sure if you want the unit tests to also involve directly making an `executeQuery` call to Spanner to read back the values and directly compare then back again to the original values?

File: spanner/cloud-client/src/main/java/com/example/spanner/LoadCsvExample.java [hunk 6]
```diff
@@ -0,0 +1,271 @@
...
+  public static final String EXCEL = "EXCEL";
+  public static final String POSTGRESQL_CSV = "POSTGRESQL_CSV";
+  public static final String POSTGRESQL_TEXT = "POSTGRESQL_TEXT}";
+
+  enum SpannerDataTypes {
+    STRING,
+    INT64,
+    FLOAT64,
+    BOOL,
+    DATE,
```
Comment (tinaspark) [inline spanner/cloud-client/src/main/java/com/example/spanner/LoadCsvExample.java:55]: We aren't supporting Date right now, but we are supporting Bytes. Can you remove DATE and add BYTES?

File: spanner/cloud-client/src/main/java/com/example/spanner/LoadCsvExample.java [hunk 7]
```diff
@@ -0,0 +1,271 @@
...
+      if (!tableColumns.containsKey(csvHeader)) {
+        System.err.println(
+            "File header " + csvHeader + " does not match any database table column name ");
+        return false;
+      }
+    }
+    return true;
+  }
+
+  /** Initialize CSV Praser format based on user specified option flags **/
```
Comment (tinaspark) [inline spanner/cloud-client/src/main/java/com/example/spanner/LoadCsvExample.java]: Parser

File: spanner/cloud-client/src/main/java/com/example/spanner/LoadCsvExample.java [hunk 8]
```diff
@@ -0,0 +1,271 @@
...
+      Reader in = new FileReader(filepath);
+      CSVFormat parseFormat = setFormat(cmd);
+      CSVParser parser = CSVParser.parse(in, parseFormat);
+
+      // If file has header, verify that header fields are valid
+      if (hasHeader && !isValidHeader(parser)) {
+        return;
+      }
+
+      // Write CVS record data to Cloud Spanner
```
Comment (tinaspark) [inline spanner/cloud-client/src/main/java/com/example/spanner/LoadCsvExample.java]: CSV

**Review by tinaspark — CHANGES_REQUESTED** [review]

File: spanner/cloud-client/src/main/java/com/example/spanner/LoadCsvExample.java (hunk 6 above)
Comment (kaylaleung) [reply spanner/cloud-client/src/main/java/com/example/spanner/LoadCsvExample.java:55]: According to the starter project documentation, I thought we are supporting Date and not Bytes. Not sure if I'm missing something of if this has changed since the starter project write up was written, but I can update if indeed we want Bytes instead of Dates?

Comment (tinaspark) [reply spanner/cloud-client/src/main/java/com/example/spanner/LoadCsvExample.java:55]: since Diem is out today, let's just go ahead and include all the Spanner types, as you have done

File: spanner/cloud-client/src/test/java/com/example/spanner/LoadCsvExampleIT.java [hunk 9]
```diff
@@ -0,0 +1,128 @@
...
+    String out;
+
+    String noHeaderPath = "src/test/resources/noHeader.csv";
+    String[] testNoHeadersArgs = new String[] {
+        instanceId, databaseId, tableName, noHeaderPath,
+    };
+    LoadCsvExample.main(testNoHeadersArgs);
+    out = bout.toString();
+    assertThat(out).contains("Data successfully written into table.");
+
```
Comment (tinaspark) [inline spanner/cloud-client/src/test/java/com/example/spanner/LoadCsvExampleIT.java:79]: can you break these into separate test methods rather than one large test for multiple files?

File: spanner/cloud-client/src/test/java/com/example/spanner/LoadCsvExampleIT.java [hunk 10]
```diff
@@ -0,0 +1,128 @@
...
+@SuppressWarnings("checkstyle:abbreviationaswordinname")
+public class LoadCsvExampleIT {
+  private String instanceId = System.getProperty("spanner.test.instance");
+  private String databaseId = System.getProperty("spanner.test.database");
+  private String tableName = System.getProperty("spanner.test.table");
+
+  static Spanner spanner;
+  static DatabaseAdminClient dbClient;
+  private ByteArrayOutputStream bout;
+  private PrintStream stdOut = System.out;
```
Comment (tinaspark) [inline spanner/cloud-client/src/test/java/com/example/spanner/LoadCsvExampleIT.java]: do you need a member variable for this? I think you could just directly use System.out in tearDown

File: spanner/cloud-client/src/test/java/com/example/spanner/LoadCsvExampleIT.java [hunk 11]
```diff
@@ -0,0 +1,128 @@
...
+  private PrintStream out;
```
Comment (tinaspark) [inline spanner/cloud-client/src/test/java/com/example/spanner/LoadCsvExampleIT.java]: I think you don't need this to be a member variable either, and it's confusing to have it here and a separate local variable inside of the test. From what I can see, the test method needs access to bout and I'm guessing that spanner and dbClient are member variables to tie their lifetime to the test class instead of the setUp function? But stdOut and out don't need to be member variables

**Review by tinaspark — CHANGES_REQUESTED** [review]

**Review by tinaspark — APPROVED** [review]
LGTM, might want to wait for Diem to approve as well

File: spanner/cloud-client/src/main/java/com/example/spanner/LoadCsvExample.java [hunk 12]
```diff
@@ -40,31 +42,40 @@
...
       return SpannerDataTypes.STRING;
-    } else if (columnType.equalsIgnoreCase("FLOAT64")) {
-      return SpannerDataTypes.FLOAT64;
     } else if (columnType.equalsIgnoreCase("INT64")) {
       return SpannerDataTypes.INT64;
+    } else if (columnType.equalsIgnoreCase("FLOAT64")) {
+      return SpannerDataTypes.FLOAT64;
+    } else if (columnType.equalsIgnoreCase("BYTE")) {
+      return SpannerDataTypes.BYTE;
+    } else if (columnType.equalsIgnoreCase("DOUBLE")) {
```
Comment (dttran) [inline spanner/cloud-client/src/main/java/com/example/spanner/LoadCsvExample.java]: I'm seeing both DOUBLE and FLOAT64 here, have you checked which types are supported? They are equivalent, I'm surprised to see both.

File: spanner/cloud-client/src/main/java/com/example/spanner/LoadCsvExample.java [hunk 13]
```diff
@@ -40,31 +42,40 @@
...
+      return SpannerDataTypes.DOUBLE;
+    } else if (columnType.equalsIgnoreCase("LONG")) {
+      return SpannerDataTypes.LONG;
```
Comment (dttran) [inline spanner/cloud-client/src/main/java/com/example/spanner/LoadCsvExample.java]: Same for LONG and INT64

File: spanner/cloud-client/src/main/java/com/example/spanner/LoadCsvExample.java [hunk 14]
```diff
@@ -213,38 +253,35 @@ public static void main(String... args) throws Exception {
...
+      // Initialize connection to Cloud Spanner
       connection = DriverManager.getConnection(
           String.format(
               "jdbc:cloudspanner:/projects/%s/instances/%s/databases/%s",
               projectId, instanceId, databaseId));
 
-      parseTableTypes(tableId);
-      parseTableHeaders(tableId);
-
+      parseTableColumns(tableName);
```
Comment (dttran) [inline spanner/cloud-client/src/main/java/com/example/spanner/LoadCsvExample.java:253]: parseTableColumns() is put into a try-catch block, but it doesn't throw in failure cases, am I missing something here?

File: spanner/cloud-client/src/main/java/com/example/spanner/LoadCsvExample.java (hunk 4 above)
Comment (dttran) [reply spanner/cloud-client/src/main/java/com/example/spanner/LoadCsvExample.java:171]: Makes sense, thanks

File: spanner/cloud-client/src/main/java/com/example/spanner/LoadCsvExample.java [hunk 15]
```diff
@@ -127,84 +134,117 @@ public static CSVFormat setFormat(CommandLine cmd) {
...
+            case STRING:
+              builder.set(columnName).to(recordValue);
               break;
             case INT64:
-              builder.set(headers.get(i)).to(Integer.parseInt(record.get(i).trim()));
+              builder.set(columnName).to(Integer.parseInt(recordValue));
               break;
             case FLOAT64:
-              builder.set(headers.get(i)).to(Float.parseFloat(record.get(i).trim()));
+              builder.set(columnName).to(Float.parseFloat(recordValue));
```
Comment (dttran) [inline spanner/cloud-client/src/main/java/com/example/spanner/LoadCsvExample.java]: I think this is parsing float32, not float64.

File: spanner/cloud-client/src/main/java/com/example/spanner/LoadCsvExample.java [hunk 16]
```diff
@@ -127,84 +134,117 @@ public static CSVFormat setFormat(CommandLine cmd) {
...
+          index++;
+        }
+        if (recordValue != null) {
+          switch (columnType) {
+            case STRING:
+              builder.set(columnName).to(recordValue);
               break;
             case INT64:
-              builder.set(headers.get(i)).to(Integer.parseInt(record.get(i).trim()));
+              builder.set(columnName).to(Integer.parseInt(recordValue));
```
Comment (dttran) [inline spanner/cloud-client/src/main/java/com/example/spanner/LoadCsvExample.java]: This will parse int32 I believe.

File: spanner/cloud-client/src/test/java/com/example/spanner/LoadCsvExampleIT.java [hunk 17]
```diff
@@ -0,0 +1,145 @@
+/*
```
Comment (dttran) [inline spanner/cloud-client/src/test/java/com/example/spanner/LoadCsvExampleIT.java:1]: Could you add some tests with the CSV files here: https://github.com/cloudspannerecosystem/sampledb?
Also look at the Dataflow job to see if it has any sample CSV.

File: spanner/cloud-client/src/main/java/com/example/spanner/LoadCsvExample.java (hunk 12 above)
Comment (kaylaleung) [reply spanner/cloud-client/src/main/java/com/example/spanner/LoadCsvExample.java]: I think Tina and I started a conversation about this in the Starter Project design doc if you wanted to continue the discussion over there.

File: spanner/cloud-client/src/main/java/com/example/spanner/LoadCsvExample.java (hunk 14 above)
Comment (kaylaleung) [reply spanner/cloud-client/src/main/java/com/example/spanner/LoadCsvExample.java:253]: `parseTableColumns` I believe throws an  SQLException if the SELECT query for information schema data fails. The try block also wraps the FileReader and the CSVParser which can also throw IOExceptions.

File: spanner/cloud-client/src/test/java/com/example/spanner/LoadCsvExampleIT.java (hunk 17 above)
Comment (kaylaleung) [reply spanner/cloud-client/src/test/java/com/example/spanner/LoadCsvExampleIT.java:1]: Added tests for a subset of the data in those CSV files since the original files exceed 20000 mutations

File: spanner/cloud-client/src/main/java/com/example/spanner/LoadCsvExample.java [hunk 18]
```diff
@@ -0,0 +1,274 @@
...
+    DATE,
+    TIMESTAMP
+  }
+
+  public static Boolean hasHeader = false;
+  public static Connection connection;
+  public static Map<String, SpannerDataTypes> tableColumns = new LinkedHashMap<>();
+
+  /** Return the data type of the column type **/
+  public static SpannerDataTypes parseSpannerDataType(String columnType) {
```
Comment (voulg) [inline spanner/cloud-client/src/main/java/com/example/spanner/LoadCsvExample.java:64]: A few related questions for this method:
1. I think it's BYTES instead of BYTE, also it is expected to have length, see here: https://cloud.google.com/spanner/docs/data-definition-language#ddl_syntax
2. All checks are case-insensitive, but the STRING regex is case sensitive. Is this intentional?

File: spanner/cloud-client/src/main/java/com/example/spanner/LoadCsvExample.java [hunk 19]
```diff
@@ -0,0 +1,274 @@
...
+      }
+    }
+    return true;
+  }
+
+  /** Initialize CSV Parser format based on user specified option flags **/
+  public static CSVFormat setFormat(CommandLine cmd) {
+    CSVFormat parseFormat;
+    // Set file format type
+    if (cmd.hasOption("f")) {
```
Comment (voulg) [inline spanner/cloud-client/src/main/java/com/example/spanner/LoadCsvExample.java:113]: Optional / out of curiosity:
Why do we support this subset of formats?

If we supported all formats could we do something like:
CSVFormat.valueOf(cmd.getOptionValue("f")) ?

File: spanner/cloud-client/src/main/java/com/example/spanner/LoadCsvExample.java [hunk 20]
```diff
@@ -0,0 +1,274 @@
...
+    STRING,
+    INT64,
+    FLOAT64,
+    BYTE,
+    BOOL,
+    DATE,
+    TIMESTAMP
+  }
+
+  public static Boolean hasHeader = false;
```
Comment (voulg) [inline spanner/cloud-client/src/main/java/com/example/spanner/LoadCsvExample.java:59]: High-level / Optional:
I understand this is just an example, but static non constant state is considered an anti-patter. For more details take a look here: go/java-practices/static#non_constants

File: spanner/cloud-client/src/main/java/com/example/spanner/LoadCsvExample.java [hunk 21]
```diff
@@ -0,0 +1,274 @@
...
+    spannerConnection.close();
+    System.out.println("Data successfully written into table.");
+  }
+
+  public static void main(String... args) throws Exception {
+
+    // Initialize option flags
+    Options opt = new Options();
+    opt.addOption("h", true, "File Contains Header");
+    opt.addOption("f", true, "Format Type of Input File");
```
Comment (voulg) [inline spanner/cloud-client/src/main/java/com/example/spanner/LoadCsvExample.java:226]: Optional: Perhaps explain what are valid formats.

File: spanner/cloud-client/src/main/java/com/example/spanner/LoadCsvExample.java [hunk 22]
```diff
@@ -0,0 +1,274 @@
...
+
+  /** Write CSV file data to Spanner using JDBC Mutation API **/
+  public static void writeToSpanner(Iterable<CSVRecord> records, String tableName, CommandLine cmd)
+      throws SQLException {
+    System.out.println("Writing data into table...");
+    List<Mutation> mutations = new ArrayList<>();
+    for (CSVRecord record : records) {
+      int index = 0;
+      WriteBuilder builder = Mutation.newInsertOrUpdateBuilder(tableName);
+      for (String columnName : tableColumns.keySet()) {
```
Comment (voulg) [inline spanner/cloud-client/src/main/java/com/example/spanner/LoadCsvExample.java:175]: Optional:
Consider adding a comment here about column ordering and why it's important that tableColumns is a LinkedHashMap.

File: spanner/cloud-client/src/main/java/com/example/spanner/LoadCsvExample.java [hunk 23]
```diff
@@ -0,0 +1,274 @@
...
+import org.apache.commons.csv.CSVParser;
+import org.apache.commons.csv.CSVRecord;
+
+/** Sample showing how to load CSV file data into Spanner */
+public class LoadCsvExample {
+  public static final String EXCEL = "EXCEL";
+  public static final String POSTGRESQL_CSV = "POSTGRESQL_CSV";
+  public static final String POSTGRESQL_TEXT = "POSTGRESQL_TEXT}";
+
+  enum SpannerDataTypes {
```
Comment (voulg) [inline spanner/cloud-client/src/main/java/com/example/spanner/LoadCsvExample.java:49]: Usually enums are not in plural, perhaps rename to SpannerDataType

File: spanner/cloud-client/src/test/resources/header.csv [hunk 24]
```diff
@@ -0,0 +1,3 @@
+int,float,bool,string,date,timestamp
```
Comment (voulg) [inline spanner/cloud-client/src/test/resources/header.csv:1]: I'm a bit confused, I thought headers are column names, but here they seem they are types?

**Review by voulg — COMMENTED** [review]
Thanks Kayla, looks very good I left a few comments.

Also you can add https://github.com/Vizerai in future changes, especially when we start touching PgAdater.

File: spanner/cloud-client/src/main/java/com/example/spanner/LoadCsvExample.java (hunk 19 above)
Comment (kaylaleung) [reply spanner/cloud-client/src/main/java/com/example/spanner/LoadCsvExample.java:113]: I believe these were just the formats specified in the initial design doc - EXCEL being assumed to be one of the more popular CSV formats and otherwise PostgreSQL CSV / TEXT being used here because they will be reused again in the main PGAdapter project as well.

File: spanner/cloud-client/src/test/resources/header.csv (hunk 24 above)
Comment (kaylaleung) [reply spanner/cloud-client/src/test/resources/header.csv:1]: Sorry, yeah this is somewhat confusing. The test case here to test the different supported types used the column names as type names - will be modified.

**Comment by kaylaleung** [comment]
Merged in by https://github.com/GoogleCloudPlatform/java-docs-samples/pull/5306
